sympy/polys/polytools.py

```diff
@@ -5279,10 +5278,11 @@ def gcd(f, g=None, *gens, **args):
 
         # gcd for domain Q[irrational] (purely algebraic irrational)
         a, b = map(sympify, (f, g))
-        if a.is_algebraic and a.is_irrational and b.is_algebraic and b.is_irrational:
-            frc = (a/b).ratsimp()
-            if frc.is_rational:
-                return a/frc.as_numer_denom()[0]
+        if isinstance(a, Expr) and isinstance(b, Expr):
```

Comment:
Is the sympify() line above supposed to ensure this? Otherwise if this code used to work on some polys objects we should add is_algebraic and is_irrational to them.

Reply:
`Poly.is_algebraic/irrational` are always None on master.
```julia
In [1]: print(Poly(1, x).is_irrational)                                                                                                                       
None
```
In this PR `Poly` does not have those attributes.

The use of `as_expr` here ensures that we get an `Expr` whether the input was `Expr` or `Poly`. It would be a useful bugfix regardless of the other changes here.

Note that I didn't write this expecting to merge this PR (yet). I don't think that the old assumptions should be on Basic rather than Expr but this is more of an experiment. However now that the tests have passed I can say that every example where code needed to be changed here shows a bug in sympy which leads to two conclusions:

1. Something similar to each of the fixes here should be applied to fix bugs on master.

2. Having the old assumptions on Basic hides bugs. They should be moved to Expr only.

Given those maybe I should flesh this PR out to something mergeable but I haven't discussed with anyone the idea of moving the old assumptions yet.

Reply:
Is the None thing also true for the other classes? It looks like adding `isinstance(expr, Expr)` to all the conditions would change the logic, but if they assumptions would be None otherwise maybe the logic hasn't changed after all.

Reply:
It doesn't change the logic because these properties are not meaningful for most non-Expr classes and so are already None in almost all cases.

Poly now subclasses Basic and never did have any `_eval_is` methods that would be needed to make these non-None:
```console
$ git grep '_eval_is' sympy/polys/
sympy/polys/rootoftools.py:    def _eval_is_real(self):
sympy/polys/rootoftools.py:    def _eval_is_imaginary(self):
```
Those are methods of `CRootOf` (which is `Expr`). There are none of these on `Poly` or any of the other classes in `polys`.

I made some changes to `ManagedProperties` to print out which non-`Expr` classes defined any of the assumptions related properties as static attributes and the only ones were cases of `is_commutative = True` (which are removed in this PR). I haven't systematically checked for `_eval_is` methods in non-`Expr` `Basic` subclasses: there aren't any that are tested as this PR shows.